### PR TITLE
Added support for PsaHashCompute to CryptoAuthLib provider.

### PR DESCRIPTION
**Pull request by RobertDrazkowskiGL**
The callback implementation is accompanied by required infrastructure enhancements, making possible provider instantiation using a real ATECCx08 hardware.
Known limitation: ATECCx08 family supports SHA256 hash only.

Signed-off-by: Robert Drazkowski <Robert.Drazkowski@globallogic.com>

### CODE REVIEW

**Review by ionut-arm — COMMENTED** [review]
Looks good to me, overall, just a couple of comments below

File: config.toml [hunk 1]
```diff
@@ -142,9 +142,25 @@ key_info_manager = "on-disk-manager"
 # provided, the value is considered to be a string.
 #owner_hierarchy_auth = "password"
 
-# Example of a CryptoAuthLib provider configuration
-# All below parameters depend on what devices, interfaces or parameters are required or supported by 
+[# Example of a CryptoAuthLib provider configuration
```
[the remark this answers is not in the document]
Comment (RobertDrazkowskiGL) [reply config.toml]: Fixed by this commit: 4a78bbe1f92ff83b59f5cc415b02c5e6b36c63a3.

File: src/providers/cryptoauthlib/mod.rs [hunk 2]
```diff
@@ -11,25 +11,38 @@ use log::trace;
...
 pub struct Provider {
-    // device: rust_cryptoauthlib::AtcaDevice,
+    device: rust_cryptoauthlib::AtcaDevice,
 }
 
 impl Provider {
     /// Creates and initialise a new instance of CryptoAuthLibProvider
-    fn new(_key_info_store: Arc<RwLock<dyn ManageKeyInfo + Send + Sync>>) -> Option<Provider> {
-        Some(Provider {})
+    // TODO - remove "pub" below. Implement ProviderBuilder.
```
[the remark this answers is not in the document]
Comment (RobertDrazkowskiGL) [reply src/providers/cryptoauthlib/mod.rs]: Definitely should not be there...
Fixed by this commit: 2ef797e23344771235f04b29973a66014e9a53b3

File: src/providers/cryptoauthlib/hash.rs [hunk 3]
```diff
@@ -0,0 +1,33 @@
+// Copyright 2020 Contributors to the Parsec project.
```
Comment (ionut-arm) [inline src/providers/cryptoauthlib/hash.rs:1]:
```suggestion
// Copyright 2021 Contributors to the Parsec project.
```

File: src/providers/cryptoauthlib/hash.rs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+use parsec_interface::operations::psa_hash_compute;
+use parsec_interface::requests::{ResponseStatus, Result};
+
+impl Provider {
+    pub(super) fn psa_hash_compute_internal(
+        &self,
+        op: psa_hash_compute::Operation,
+    ) -> Result<psa_hash_compute::Result> {
+        let mut hash = vec![0u8; op.alg.hash_length()];
+        let message = op.input.to_vec();
```
Comment (ionut-arm) [inline src/providers/cryptoauthlib/hash.rs:14]: Could you make sure that this buffer is zeroized if not sent to the `cryptoauthlib` library (i.e. in the `_` branch of the `match` below) ?

**Review by ionut-arm — COMMENTED** [review]
Thanks for the fixes!

File: src/providers/cryptoauthlib/hash.rs (hunk 4 above)
Comment (RobertDrazkowskiGL) [reply src/providers/cryptoauthlib/hash.rs:14]: Forgive me detailed question - you suggest to zeroize hash (the output) when algorithm is not supported by CAL/ATECC?

Comment (ionut-arm) [reply src/providers/cryptoauthlib/hash.rs:14]: No - to zeroize the `message` that you create on this line

File: Cargo.toml [hunk 5]
```diff
@@ -46,6 +46,7 @@ libc = "0.2.77"
 anyhow = "1.0.32"
 # Using a fork until the JWT support is merged into the main rust-spiffe repository
 spiffe = { git = "https://github.com/hug-dev/rust-spiffe", branch = "refactor-jwt" }
+rust-cryptoauthlib = { git = "https://github.com/RobertDrazkowskiGL/rust-cryptoauthlib.git", optional = true }
```
[the remark this answers is not in the document]
Comment (ionut-arm) [reply Cargo.toml:49]: Sorry, forgot to update this - we're not in a hurry to release a new version to crates.io and you already added features to "hide" the new provider, so it's all good!

File: src/providers/cryptoauthlib/hash.rs (hunk 4 above)
Comment (RobertDrazkowskiGL) [reply src/providers/cryptoauthlib/hash.rs]: I admit I had an aesthetic problem with this zeroizing (or maybe the way I applied this?). I reduced the scope of the variable to a block where it is actually useful. Hope this addresses your issue as well.

Comment (ionut-arm) [reply src/providers/cryptoauthlib/hash.rs]: That works!

Just for future reference, what I had in mind was importing `zeroize::Zeroize` in the file and then calling [`message.zeroize()`](https://docs.rs/zeroize/1.2.0/zeroize/trait.Zeroize.html#tymethod.zeroize) in that `match` branch - is this what you tried?

Comment (RobertDrazkowskiGL) [reply src/providers/cryptoauthlib/hash.rs]: Yes, this was my first attempt. And that `message.zeroize()` looked for me as a workaround while narrowing down the scope fixed the issue by design.

**Comment by RobertDrazkowskiGL** [comment]
I am not able to reproduce this error in "Continuous Integration / Integration tests using PKCS 11 provider". 
Is there a way to restart CI without submitting a code?

**Comment by ionut-arm** [comment]
> 
> 
> I am not able to reproduce this error in "Continuous Integration / Integration tests using PKCS 11 provider".
> Is there a way to restart CI without submitting a code?

Don't worry about that error, it's something that seems to be failing intermittently most likely because of the library we use for simulating a PKCS11 backend. We need to figure out the best way to keep the tests going, but for now we just ignore the failure (especially since you don't touch any part of the PKCS11 provider)

**Review by ionut-arm — APPROVED** [review]
Looks good!

**Review by hug-dev — APPROVED** [review]
Thanks!!

**Pull request merged**